src/main.rs

```diff
@@ -1,13 +1,317 @@
+use memory_management::memory_read;
+use operations::*;
+use std::io::{Read, stdin};
+use std::ops::{Index, IndexMut};
+use std::time::Duration;
+use std::{env, io};
+use termios::*;
+pub mod file_management;
+mod operations;
+mod tests;
+use std::os::fd::AsRawFd;
+use timeout_readwrite::TimeoutReadExt;
+mod memory_management;
+
+static MEM_MAX: usize = 1 << 16;
+static PC_START: u16 = 0x3000;
+/// Traps are predefined routines, each trap in the enum represents a routine
+pub enum Traps {
+    Getc = 0x20,
+    Out = 0x21,
+    Puts = 0x22,
+    In = 0x23,
+    Putsp = 0x24,
+    Halt = 0x25,
+}
+
+impl Traps {
+    fn from(value: u16) -> Result<Traps, Errors> {
+        match value {
+            0x20 => Ok(Traps::Getc),
+            0x21 => Ok(Traps::Out),
+            0x22 => Ok(Traps::Puts),
+            0x23 => Ok(Traps::In),
+            0x24 => Ok(Traps::Putsp),
+            0x25 => Ok(Traps::Halt),
+            badcode => Err(Errors::BadTrapCode(badcode.to_string())),
+        }
+    }
+}
+pub enum Errors {
+    BadRegisterReference(String),
+    BadOpCode(String),
+    BadFile(String),
+    DisableInputBuffering,
+    RestoreInputBuffering,
+    BadTrapCode(String),
+    Trap(Traps),
+    FewArguments,
+    BadTermios,
+}
+#[derive(Clone, Copy)]
+enum Registers {
+    Rr0,
+    Rr1,
+    Rr2,
+    Rr3,
+    Rr4,
+    Rr5,
+    Rr6,
+    Rr7,
+    Rpc,
+    Rcond,
+    Rcount,
+}
```

Comment:
There's a convention in Rust to not prefix enum variants.

Here's a suggestion with the prefix removed and two registers renamed, assuming the `Rcond` and `Rcount` are used for conditional flags and instruction counter respectively. The `R` in `R0`, `R1`... is still there because it's part of the register name, but not a prefix of the enum variants.
```suggestion
enum Registers {
    R0,
    R1,
    R2,
    R3,
    R4,
    R5,
    R6,
    R7,
    Pc,
    Flags,
    InstRet,
}
```

Reply:
Done